apps/system/flash_test/flash_test.c

```diff
@@ -59,80 +59,553 @@
 #include <errno.h>
 #include <stdint.h>
 #include <sys/types.h>
-
+#include <sys/time.h>
+#include <arch/board/board.h>
+#include <tinyara/spi/spi.h>
 #include <tinyara/fs/mtd.h>
-
+#include <time.h>
+#include <tinyara/config.h>
+#include <tinyara/fs/ioctl.h>
+#include <fcntl.h>
 /****************************************************************************
  * Definitions
  ****************************************************************************/
-#define BUF_SIZE   1024
-
+#define BUF_SIZE   65536
+#define SIZE_1MB   1024*1024 
 #define READ_CMD    "read"
+#define WRITE_CMD   "write"
+#define ERASE_CMD   "erase"
+#define BCDEV_CMD	"bch"
+
+#ifdef CONFIG_BUILD_FLAT
+#define INT_FLASH_PART 7
+#define EXT_FLASH_PART 13
+#else
+#define INT_FLASH_PART 11
+#define EXT_FLASH_PART 17
+#endif
+
+
 /****************************************************************************
  * Private Data
  ****************************************************************************/
+enum flash_e {
+	FLASH_INTERNAL = 0,
+	FLASH_EXT = 1
+};
+typedef enum flash_e flash_t;
+
+static FAR struct mtd_geometry_s geo;
+
+static long sys_start_time = 0;
+static long sys_end_time = 0;
+static long systick_time_ms = 0;
+
+static char blockname[32];
+
+static inline long get_time(void)
+{
+	struct timeval tv;
+
+	gettimeofday(&tv, NULL);
+	return (tv.tv_sec * 1000 + tv.tv_usec / 1000);
+}
+
+static int prepare_bch(flash_t flash)
+{
+	int ret;
+	int fd;
+	int g_partno = 0;
+
+	if (flash == FLASH_INTERNAL) {
+		g_partno = INT_FLASH_PART;
+	} else if (flash == FLASH_EXT) {
+		g_partno = EXT_FLASH_PART;
+	}
+	snprintf(blockname, 32, "/dev/mtdblock%d", g_partno);
+
+	fd = open(blockname, O_RDWR);
+	if (fd < 0) {
+		printf("ERROR: open bch %s\n", blockname);
+		fflush(stdout);
+		return ERROR;
+	}
+	ret = ioctl(fd, MTDIOC_GEOMETRY, (unsigned long)&geo);
+	if (ret < 0) {
+		printf("ERROR : Not able to get the BCH Flash Geometry:\n");
+		close(fd);
+		return ERROR;
+	}
+	printf("BCH Flash Geometry:\n");
+	printf("  blocksize:      %lu\n", (unsigned long)geo.blocksize);
+	printf("  erasesize:      %lu\n", (unsigned long)geo.erasesize);
+	printf("  neraseblocks:   %lu\n", (unsigned long)geo.neraseblocks);
+	close(fd);
+	return OK;
+}
+
+static void fill_buffer(FAR uint32_t *buff)
+{
+	off_t offset = 0;
+	for (int i = 0; i < BUF_SIZE / sizeof(uint32_t); i++) {
+		buff[i] = offset;
+		offset += 4;
+	}
+}
+
+static int bch_write(flash_t flash)
+{
+	FAR uint32_t *buffer;
+	int fd;
+	ssize_t nbytes;
+	off_t offset;
+	int g_partno = 0;
+
+	if (flash == FLASH_INTERNAL) {
+		g_partno = INT_FLASH_PART;
+	} else if (flash == FLASH_EXT) {
+		g_partno = EXT_FLASH_PART;
+	}
+	/* Allocate a buffer */
+	buffer = (FAR uint32_t *)malloc(BUF_SIZE);
+
+	if (!buffer) {
+		printf("ERROR: failed to allocate a sector buffer\n");
+		fflush(stdout);
```

Comment:
Is this necessary?

Reply:
not sure. extended the existing usage. anyway removed this.